services/course-material/src/components/chatbot/MessageBubble.tsx

```diff
@@ -58,33 +121,315 @@ const messageStyle = css`
+  border: none;
+  margin: 0 0.5rem;
+  color: ${baseTheme.colors.gray[400]};
+  transition: filter 0.2s;
+
+  &:hover {
+    filter: brightness(0.9) contrast(1.1);
+  }
+`
+
+const referenceListStyle = (expanded: boolean, referenceList: string) => css`
+  display: flex;
+  ${expanded ? `flex-flow: column nowrap;` : `flex-flow: row nowrap;`}
+
+  #${referenceList} {
+    display: flex;
+    flex: 10;
+    ${expanded
+      ? `
+    flex-flow: column nowrap;
+    padding: 7px;
+    justify-content: space-around;
+    `
+      : `
+    flex-flow: row nowrap;
+    overflow: hidden;
+    white-space: pre;
+    mask-image: linear-gradient(0.25turn, black 66%, transparent);
+  `}
+  }
+`
 
 let md = getRemarkable()
 
 const MessageBubble: React.FC<MessageBubbleProps> = ({
   message,
   isFromChatbot,
   isPending,
-  hideCitations,
+  citations,
 }) => {
-  const processedMessage = useMemo(() => {
-    let renderedMessage = message
-    if (hideCitations) {
-      renderedMessage = renderedMessage.replace(/\[.*?\]/g, "")
-    }
+  const { t } = useTranslation()
+
+  const referenceListId = useId()
+  let triggerRef = useRef<HTMLElement>(null)
+
+  const [citationsOpen, setCitationsOpen] = useState(false)
+  const [popoverState, dispatch] = useReducer(popperStateReducer, {
+    refElement: null,
+    citationButtonClicked: false,
+  })
+
+  let { hoverProps: hoverCitationProps, isHovered: isCitationHovered } = useHover({
+    onHoverStart: (e) => {
+      dispatch({ type: "HOVER_CITATION_BUTTON", payload: e.target })
+    },
+    onHoverEnd: (e) => {
+      dispatch({ type: "UNHOVER_CITATION_BUTTON", payload: e.target })
+    },
+  })
+  let { hoverProps: hoverPopoverProps, isHovered: isPopoverHovered } = useHover({})
+
+  useEffect(() => {
+    triggerRef.current = popoverState.refElement
+  }, [popoverState.refElement])
+
+  const [processedMessage, processedCitations, citationTitleLen] = useMemo(() => {
+    let renderedMessage: ReactElement[] = []
+    let filteredCitations: ChatbotConversationMessageCitation[] = []
+
     if (isFromChatbot) {
-      renderedMessage = sanitizeCourseMaterialHtml(md.render(renderedMessage).trim())
+      let messageCopy = message
+      messageCopy = sanitizeCourseMaterialHtml(md.render(messageCopy).trim()).slice(3, -4)
+      // slicing to remove the <p>-tags that envelope the whole message, since they will be broken
+      // when the message is split into parts later
+      let citedDocs = Array.from(messageCopy.matchAll(/\[[a-z]*?([0-9]+)\]/g), (arr, _) =>
+        parseInt(arr[1]),
+      )
+      let citedDocsSet = new Set(citedDocs)
+      filteredCitations = citations
+        ? citations.filter((cit) => citedDocsSet.has(cit.citation_number))
+        : []
+      if (citationsOpen && filteredCitations.length > 0) {
+        // if there are citations in text, render buttons for them & md
+        let messageParts = messageCopy.split(/\[[a-z]*?[0-9]+\]/g)
+        renderedMessage = messageParts.map((s, i) => {
```

Comment:
Own component for rendering these so that we don't have to construct an array with React elements to be rendered later.